internal/provider/provider.go

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/vault/api"
+
+	"github.com/hashicorp/terraform-provider-vault/internal/consts"
+)
+
+const DefaultMaxHTTPRetriesCCC = 10
+
+// Description is essentially a DataSource or Resource with some additional metadata
+// that helps with maintaining the Terraform Vault Provider.
+type Description struct {
+	// PathInventory is used for taking an inventory of the supported endpoints in the
+	// Terraform Vault Provider and comparing them to the endpoints noted as available in
+	// Vault's OpenAPI description. A list of Vault's endpoints can be obtained by,
+	// from Vault's home directory, running "$ ./scripts/gen_openapi.sh", and then by
+	// drilling into the paths with "$ cat openapi.json | jq ".paths" | jq 'keys[]'".
+	// Here's a short example of how paths and their path variables should be represented:
+	//		"/transit/keys/{name}/config"
+	//		"/transit/random"
+	//		"/transit/random/{urlbytes}"
+	//		"/transit/sign/{name}/{urlalgorithm}"
+	PathInventory []string
+
+	// EnterpriseOnly defaults to false, but should be marked true if a resource is enterprise only.
+	EnterpriseOnly bool
+
+	Resource *schema.Resource
+}
+
+type ResourceRegistry map[string]*Description
+
+func NewProvider(
+	dataRegistry ResourceRegistry,
+	resourceRegistry ResourceRegistry,
+	resourceMap ...map[string]*schema.Resource,
+) *schema.Provider {
+	dataSourcesMap, err := parse(dataRegistry)
+	if err != nil {
+		panic(err)
+	}
+
+	resourcesMap, err := parse(resourceRegistry)
+	if err != nil {
+		panic(err)
+	}
+
+	for _, m := range resourceMap {
+		MustAddSchemaResource(m, resourcesMap, nil)
```

Comment:
should this be
```suggestion
		MustAddSchemaResource(m, resourceMap, nil)
```

Reply:
Variable names a bit confusing, I think what we have there is correct. It comes from:
https://github.com/hashicorp/terraform-provider-vault/pull/1884/files#diff-7e69de666c9006fa0a0fb8fe4e4f8c418acce7c555f69f8087d88d16b434f2c1L59

Reply:
Ah you're right, thanks for clarifying

Reply:
Make the variable names a little less confusing: f663882da8043ae07f30111e8b206349b1e2b592

Reply:
Nice!